validator/index.js

```diff
@@ -25,50 +25,226 @@ const path = require('path');
+     * 'ERROR', and 'WARNING'.
+     */
+    this.severity = 'UNKNOWN_SEVERITY';
+    /**
+     * The line number at which the error was seen (1 is the first line).
+     */
+    this.line = 1;
+    /**
+     * The column number at which the error was seen (0 is the first column).
+     */
+    this.col = 0;
+    /**
+     * A human-readable error message for the validation error.
+     * If you find yourself trying to write a parser against this string
+     * to scrape out some detail, consider looking at the code and params
+     * fields below.
+     * @type {!string}
+     */
+    this.message = '';
+    /**
+     * The spec URL is often added by the validator to provide additional
+     * context for the error. In a user interface this would be shown
+     * as a "Learn more" link.
+     * @type {!string}
+     */
+    this.specUrl = null;
+    /**
+     * Categorizes error messages into higher-level groups. This makes it
+     * easier to create error statistics across a site and give advice based
+     * on the most common problems for a set of pages.
+     * See the ErrorCategory.Code enum in validator.proto for possible values.
+     * @type {!string}
+     */
+    this.category = 'UNKNOWN';
+    /**
+     * This field is only useful when scripting against the validator,
+     * it should not be displayed in a user interface as it adds nothing
+     * for humans to read over the message field (see above).
+     * Possible values are the codes listed in ValidationError.Code - see
+     * validator.proto. Examples: 'UNKNOWN_CODE', 'MANDATORY_TAG_MISSING',
+     * 'TAG_REQUIRED_BY_MISSING'. For each of these codes there is a
+     * format string in validator-main.protoascii (look for error_formats),
+     * which is used to assemble the message from the strings in params.
+     * @type {!string}
+     */
+    this.code = 'UNKNOWN_CODE';
+    /**
+     * This field is only useful when scripting against the validator,
+     * it should not be displayed in a user interface as it adds nothing
+     * for humans to read over the message field (see above).
+     * @type {!Array<!string>}
+     */
+    this.params = [];
+  }
+}
+exports.ValidationError = ValidationError;
+
+/**
+ * A global static map used by the validateString function to avoid loading
+ * AMP Validators more than once.
+ * @type {!Object<string, Object>}
+ */
+const precompiledByValidatorJs = {};
+
+/**
+ * @param {!string} inputString
+ * @param {string=} validatorJs_opt
+ * @returns {Promise.<(ValidationResult|Error)>}
+ * @export
+ */
+function validateString(inputString, validatorJs_opt) {
+  const validatorJs =
+      validatorJs_opt || 'https://cdn.ampproject.org/v0/validator.js';
+  const validate = (sandbox, resolve) => {
+    const internalResult = sandbox.amp.validator.validateString(inputString);
+    const result = new ValidationResult();
+    result.status = internalResult.status;
+    for (const internalError of internalResult.errors) {
```

Comment:
Yes, this part would be easy to change, but may not buy us much by itself. I'm more after the other fancy features like promises and class and the way the vm behaves (see the other comment please).